File: readability/readability.c
#include <ctype.h>
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int get_count_letters(string txt);
int get_count_words(string txt);
int get_count_sentences(string txt);
float get_reading_level(int l, int w, int s);

int main(void)
{
    string text = get_string("Enter text to evaluate US reading level: ");

    int num_letters = get_count_letters(text);
    int num_words = get_count_words(text);
    int num_sentences = get_count_sentences(text);
    float reading_level = get_reading_level(num_letters, num_words, num_sentences);

    if (reading_level < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (reading_level > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", (int) round(reading_level));
    }
}


int get_count_letters(string txt)
{
    int k = 0, count = 0;
    while (txt[k] != '\0')
    {
        if (isalpha(txt[k]))
        {
            count++;
        }
        k++;
    }
    return count;
}


int get_count_words(string txt)
{
    int k = 0, count = 0;
    bool at_least_one_space = false;

    while (txt[k] != '\0')
    {
        if (txt[k] == ' ')
        {
            count++;
            at_least_one_space = true;
        }
        k++;
    }
    if (at_least_one_space)
    {
        count++;
        return count;
    }
    else
    {
        return 1;
    }
}


int get_count_sentences(string txt)
{
    int k = 0, count = 0;
    while (txt[k] != '\0')
    {
        if (txt[k] == '.' || txt[k] == '!' || txt[k] == '?')
        {
            count++;
        }
        k++;
    }
    return count;
}


float get_reading_level(int l, int w, int s)
{
    // Computes the Coleman-Liau index for US reading level based on the number of letters,
    // words and sentences in a text
    float index = (0.0588 * (((float) l / (float) w) * 100)) - (0.296 * (((float) s / (float) w) * 100)) - 15.8;

    return index;
}

File: sentimental-readability/readability.py
import re

def main():
    text = input("Enter text to evaluate US reading level: ").strip()

    num_letters = get_num_letters(text)
    num_words = get_num_words(text)
    num_sentences = get_num_sentences(text)
    print_reading_level(num_letters, num_words, num_sentences)

def get_num_letters(text):
    alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

    num = 0
    for letter in text:
        if letter.upper() in alphabet:
            num = num + 1
    return num

def get_num_words(text):
    return len(text.split())

def get_num_sentences(text):
    sentences = re.split(r"(?<=[.!?]) +", text)
    return len(sentences)

def print_reading_level(num_letters, num_words, num_sentences):
    index = 0.0588 * ((num_letters/num_words)*100) - 0.296 * ((num_sentences/num_words)*100) - 15.8

    if index < 1:
        print("Before Grade 1")
    elif index  > 16:
        print("Grade 16+")
    else:
        print(f"Grade {int(round(index))}")


if __name__ == "__main__":
    main()
